test/distributed/test_c10d_gloo.py

```diff
@@ -0,0 +1,2130 @@
+    for i in range(1, 5):
+        vin = reduce(operator.or_, [rank * i + j for j in range(i)])
+        vout = reduce(operator.or_, range(world_size * i))
+        tests.append(
+            (
+                c10d.ReduceOp.BOR,
+                torch.tensor([vin], dtype=torch.int32),
+                torch.tensor([vout], dtype=torch.int32),
+            ),
+        )
+
+    # Generate tests for XOR.
+    # These emulate a larger world size per iteration by having every
+    # rank contribute multiple values that are pre-XOR'ed.
+    for i in range(1, 5):
+        vin = reduce(operator.xor, [rank * i + j for j in range(i)])
+        vout = reduce(operator.xor, range(world_size * i))
+        tests.append(
+            (
+                c10d.ReduceOp.BXOR,
+                torch.tensor([vin], dtype=torch.int32),
+                torch.tensor([vout], dtype=torch.int32),
+            ),
+        )
+
+    return tests
+
+
+def simple_coalesced_reduce_tests(rank, world_size):
+    return [
+        (
+            c10d.ReduceOp.SUM,
+            [torch.tensor([rank + 1]), torch.tensor([(rank + 1) ** 2])],
+            [
+                torch.tensor([float(world_size * (world_size + 1) / 2)]),
+                torch.tensor(
+                    [float(world_size * (world_size + 1) * (2 * world_size + 1) / 6)]
+                ),
+            ],
+        ),
+        (
+            c10d.ReduceOp.PRODUCT,
+            [torch.tensor([rank + 1.0]), torch.tensor([rank + 2.0])],
+            [
+                torch.tensor([float(math.factorial(world_size))]),
+                torch.tensor([float(math.factorial(world_size + 1))]),
+            ],
+        ),
+        (
+            c10d.ReduceOp.MIN,
+            [torch.tensor([rank + x]) for x in [0.0, 1.0]],
+            [torch.tensor([0.0]), torch.tensor([1.0])],
+        ),
+        (
+            c10d.ReduceOp.MAX,
+            [torch.tensor([rank + x]) for x in [1.0, 2.0]],
+            [torch.tensor([world_size]), torch.tensor([world_size + 1.0])],
+        ),
+    ]
+
+
+def simple_multi_input_reduce_tests(rank, world_size):
+    return [
+        (
+            c10d.ReduceOp.SUM,
+            [torch.tensor([2 * rank + 0.0]), torch.tensor([2 * rank + 1.0])],
+            torch.tensor([float(world_size * (2 * world_size - 1))]),
+        ),
+        (
+            c10d.ReduceOp.PRODUCT,
+            [torch.tensor([2 * rank + 1.0]), torch.tensor([2 * rank + 2.0])],
+            torch.tensor([float(math.factorial(2 * world_size))]),
+        ),
+        (
+            c10d.ReduceOp.MIN,
+            [torch.tensor([2 * rank + 1.0]), torch.tensor([2 * rank + 2.0])],
+            torch.tensor([1.0]),
+        ),
+        (
+            c10d.ReduceOp.MAX,
+            [torch.tensor([2 * rank + 1.0]), torch.tensor([2 * rank + 2.0])],
+            torch.tensor([2 * world_size]),
+        ),
+    ]
+
+
+class RendezvousEnvTest(TestCase):
+    @retry_on_connect_failures
+    def test_logging_init(self):
```

Comment:
Why is this test in gloo rather than common?

Reply:
line 181:
```
        c10d.init_process_group(backend="gloo", init_method="env://")
```

Reply:
Better to add `@requires_gloo()` here.